src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+                    };
+                    let entry = time_data.entry((time, row)).or_insert(0);
+                    *entry += diff;
+                }
+            }
+        }
+
+        // Now let's prepare the output
+        let mut buf = Vec::new();
+        assert!(lower.is_some());
+        assert!(upper.is_some());
+
+        let lower = lower.unwrap();
+        let upper = upper.unwrap();
+
+        // Frame each batch with its lower and upper bound timestamp.
+        // TODO: match the behavior of CDCv2 updates with a count of messages
+        // at each timestamp.
+        encode_progress(lower, &mut buf)?;
+        for ((timestamp, row), diff) in time_data.into_iter() {
+            if diff == 0 {
+                continue;
+            }
+
+            encode_update(row, timestamp, diff, &mut buf)?;
+        }
+
+        encode_progress(upper, &mut buf)?;
+
+        let batch_name = format!("batch-{}-{}", lower, upper);
+        let batch_path = trace_path.join(&batch_name);
+        let batch_tmp_path = trace_path.join(format!("{}-tmp", batch_name));
+        // Write the file first suffixed with "-tmp" and then rename to guard against
+        // partial writes.
+        fs::write(&batch_tmp_path, buf)
+            .with_context(|| format!("failed to write batch file {}", batch_tmp_path.display()))?;
+        fs::rename(&batch_tmp_path, &batch_path).with_context(|| {
+            format!(
+                "failed to rename batch file from: {} to: {}",
+                batch_tmp_path.display(),
+                batch_path.display()
+            )
+        })?;
+
+        Ok(Batch {
+            upper,
+            lower,
+            path: batch_path,
+        })
+    }
+
+    /// Reintroduce a batch based on an available file in the trace
+    /// directory.
+    fn reinit(path: PathBuf) -> Result<Self, anyhow::Error> {
+        let batch_name = path
+            .file_name()
+            .expect("batch name known to exist")
+            .to_str()
+            .expect("batch name known to be valid utf8");
+        let parts: Vec<_> = batch_name.split('-').collect();
+        assert!(parts.len() == 3);
+        Ok(Self {
+            upper: parts[2].parse()?,
+            lower: parts[1].parse()?,
+            path,
+        })
+    }
+
+    /// Read the data from a batch stored on disk into memory.
+    fn read(&self) -> Result<Vec<Message>, anyhow::Error> {
+        read_segment(&self.path)
+    }
+
+    /// Physically concatenate all batches together into a single batch
+    /// and consolidate updates up to the frontier if provided.
+    fn compact(
+        batches: &[Batch],
+        trace_path: &Path,
+        compaction_frontier: Option<Timestamp>,
+    ) -> Result<Self, anyhow::Error> {
+        let mut messages = vec![];
+
+        for batch in batches {
+            messages.append(&mut read_segment(&batch.path)?);
+        }
+
+        Batch::create_from_messages(messages, trace_path, compaction_frontier)
```

Comment:
Do `batches` need to be contiguous for this to be correct? Probably, right? It seems like a fine thing to verify hereIt also seems (in analogous DD code) like a great point to explicitly state the lower and upper bounds at the batch creation, rather than to re-extract them from the progress messages themselves.

Reply:
yes and agree. added a todo